src/lib/operators/join_helper/join_output_writing.hpp

```diff
@@ -0,0 +1,269 @@
+#pragma once
+
+#include "storage/create_iterable_from_segment.hpp"
+#include "storage/segment_iterate.hpp"
+
+namespace opossum {
+
+enum class OutputColumnOrder { LeftFirstRightSecond, RightFirstLeftSecond, RightOnly };
+
+using PosLists = std::vector<std::shared_ptr<const AbstractPosList>>;
+using PosListsByChunk = std::vector<std::shared_ptr<PosLists>>;
+
+/**
+ * Returns a vector where each entry with index i references a PosLists object. The PosLists object
+ * contains the position list of every segment/chunk in column i.
+ * @param input_table
+ */
+// See usage in _on_execute() for doc.
+inline PosListsByChunk setup_pos_lists_by_chunk(const std::shared_ptr<const Table>& input_table) {
+  Assert(input_table->type() == TableType::References, "Function only works for reference tables");
+
+  std::map<PosLists, std::shared_ptr<PosLists>> shared_pos_lists_by_pos_lists;
```

Comment:
Are we sure that we don't want an `unordered_map`?

Reply:
@Bouncner I replaced the `std:.map` with `std::unordered_map`. Can you please have a look if you are happy with the implementation especially with the hasher.